File: 003 Longest Substring Without Repeating Characters/longest_substring_wo_repeating.cpp
/*	
 *	LeetCode Submissions by Xinyu Chen
 *	Longest Substring Without Repeating Characters
 *	https://leetcode.com/problems/longest-substring-without-repeating-characters/
 *	Runtime: 76 ms
 */

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_set<char> t;
        int l = s.size();
        if (l < 2) return l;
        int maxlen = 1;
        int left = 0, right = 1;
        st.insert(s[left]);
        while (right < l) {
            if (st.find(s[right]) == st.end()) {
                st.insert(s[right++]);
                maxlen = max(maxlen, right - left);
            }
            else st.erase(s[left++]);
        }
        return maxlen;
    }
};

File: 003 Longest Substring Without Repeating Characters/longest_substring_wo_repeating.java
// author: Bonnie

public class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> map = new HashMap<Character, Integer>();
        int rtn = 0;
        int left = 0;
        for(int i=0; i<s.length(); i++){    
            char c = s.charAt(i);
            if(map.get(c)!=null){
                int last = map.get(c);
                left = Math.max(left, last+1);
            }
            map.put(c, i);
            rtn = Math.max(rtn, i-left+1);
        }
        return rtn;
    }
}
